pkg/stanza/adapter/config.go

```diff
@@ -15,13 +15,43 @@
 package adapter // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/adapter"
 
 import (
+	"time"
+
 	"go.opentelemetry.io/collector/component"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
 )
 
+// RetryConfig defines configuration for retrying batches in case of receiving a retryable error from a downstream
+// consumer. The current supported strategy is exponential backoff.
+type RetryConfig struct {
+	// Enabled indicates whether to not retry sending logs in case of receiving a retryable error from a downstream
+	// consumer. Default is false.
+	Enabled bool `mapstructure:"enabled"`
+	// InitialInterval the time to wait after the first failure before retrying. Default value is 1 second.
+	InitialInterval time.Duration `mapstructure:"initial_interval"`
+	// MaxInterval is the upper bound on backoff interval. Once this value is reached the delay between
+	// consecutive retries will always be `MaxInterval`. Default value is 30 seconds.
+	MaxInterval time.Duration `mapstructure:"max_interval"`
+	// MaxElapsedTime is the maximum amount of time (including retries) spent trying to send a logs batch to
+	// a downstream consumer. Once this value is reached, the data is discarded. It never stops if MaxElapsedTime == 0.
+	// Default value is 5 minutes.
+	MaxElapsedTime time.Duration `mapstructure:"max_elapsed_time"`
+}
```

Comment:
Shall we warn and document against setting `MaxElapsedTime` to `0` ; 
My concern is that we can hit a deadlock with value `0`, especially if the amount of "free memory" needed by the memory limiter, after GC, is not met (ie: keeping `mustRefuse` set to true) unless we drop the data are trying to send.

Reply:
@dmitryax on a slightly related note, can we have a total max data the receiver can read/process at a time? The thought behind it, is to workaround situations where the receiver reads a large amount of logs, fills up the memory right away, and can't even send the first batch. We have seen this happening in production.

Reply:
> @dmitryax on a slightly related note, can we have a total max data the receiver can read/process at a time? The thought behind it, is to workaround situations where the receiver reads a large amount of logs, fills up the memory right away, and can't even send the first batch. We have seen this happening in production.

The file reader reads logs line by line and passes each of them to the downstream operators separately, it has a `max_log_size` option to avoid reading long lines, which is 1Mb by default. Then we also have the `recombine` operator, which potentially can cause issues like that when many logs are stitched together. We recently added the same 1Mb limit in our helm chart config to avoid that. I believe that should be sufficient.

Reply:
> Shall we warn and document against setting MaxElapsedTime to 0 ;
> My concern is that we can hit a deadlock with value 0, especially if the amount of "free memory" needed by the memory limiter, after GC, is not met (ie: keeping mustRefuse set to true) unless we drop the data are trying to send.

I think it's pretty unlikely that all the memory will be taken by the file log receiver. The maximum number of log records it can accumulate is 100 (the default batch size). The maximum possible size of each log is 1Mb which is very unlikely to be reached.